### PR TITLE
CAT-FIX Various fixes from feature/NOLB-branch

### PR DESCRIPTION
**Pull request by sjaindl**
These are some commits that are in feature/NOLB and should also be merged into develop

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/common/LookData.java [hunk 1]
```diff
@@ -190,16 +197,16 @@ public String getChecksum() {
 		return fileName.substring(0, 32);
 	}
 
-	protected String getPathToImageDirectory() {
+	String getPathToImageDirectory() {
 		return Utils.buildPath(Utils.buildProjectPath(ProjectManager.getInstance().getCurrentProject().getName()),
 				getSceneNameByLookData(), Constants.IMAGE_DIRECTORY);
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/common/LookData.java]: Use `Utils.buildScenePath`. It already internally encodes the project and scene names.

Eventually we should create a handful of helper methods for many scenarios. e.g. in this case `buildImagePath(Project, Scene)` and [Replace String paths by File](https://jira.catrob.at/browse/CAT-1304).

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BackPackedData.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.catrobat.catroid.formulaeditor.UserVariable;
+
+import java.io.Serializable;
+
+public class BackPackedData implements Serializable {
+	public UserVariable userVariable;
+	public Integer userVariableType;
+
+	public UserList userList;
+	public Integer userListType;
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BackPackedData.java]: Stupid question: Why does a BackPackedData store two things, a UserVariable and a UserList?

File: catroid/src/main/java/org/catrobat/catroid/common/LookData.java (hunk 1 above)
Comment (sjaindl) [inline catroid/src/main/java/org/catrobat/catroid/common/LookData.java]: Yes, I would vote for that - the helper methods should also include sounds and the backpack.

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BackPackedData.java (hunk 2 above)
Comment (sjaindl) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BackPackedData.java]: Actually, I've moved it to a central class out of the Variable- and List-Bricks.
But I admit that it makes sense to split it into 2 classes - one for variables and one for lists. I've changed that accordingly.

**Comment by sjaindl** [comment]
rebased
@robertpainsi should be fine now ;)

File: catroid/src/main/java/org/catrobat/catroid/ProjectManager.java [hunk 3]
```diff
@@ -746,20 +747,40 @@ private void correctAllNestedReferences(Script script) {
 			} else if (currentBrick instanceof LoopBeginBrick) {
 				loopBeginList.add((LoopBeginBrick) currentBrick);
 			} else if (currentBrick instanceof LoopEndBrick) {
+				if (!hasItems(loopBeginList)) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: `!hasItems(loopBeginList)` === `loopBeginList.isEmpty()`

File: catroid/src/main/java/org/catrobat/catroid/ProjectManager.java [hunk 4]
```diff
@@ -769,6 +790,18 @@ private void correctAllNestedReferences(Script script) {
...
 		}
+
+		removeBricksWithInvalidReferences(bricksWithInvalidReferences, script);
+	}
+
+	private void removeBricksWithInvalidReferences(ArrayList<Brick> bricksWithInvalidReferences, Script script) {
+		for (Brick brick : bricksWithInvalidReferences) {
+			script.removeBrick(brick);
+		}
+	}
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: I think the script itself is a better place for this method. `script.removeBricks(List bricks)`. Then you can use `script.removeBricks(bricksWithInvalidReferences)`.

Comment (sjaindl) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: Yes, that should be clearer

File: catroid/src/main/java/org/catrobat/catroid/ProjectManager.java (hunk 3 above)
Comment (sjaindl) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: changed

Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: thanks!

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 5]
```diff
@@ -146,6 +146,12 @@ public void removeInstancesOfUserBrick(UserBrick userBrickToRemove) {
 		}
 	}
 
+	public void removeBricksWithInvalidReferences(List<Brick> bricksWithInvalidReferences) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java]: Please rename to `removeBricks(List<Brick> bricks)`. So it's a more generic method.

Comment (sjaindl) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java]: true true

**Pull request merged**